docs/proposals/rescheduling.md

```diff
@@ -0,0 +1,496 @@
+achieve rescheduling, and only briefly discuss example policies. We expect that community
+experimentation will lead to a significantly better understanding of the range, potential,
+and limitations of rescheduling policies.
+
+## Example use cases
+
+Example use cases for rescheduling are
+* moving a running pod onto a node that better satisfies its scheduling criteria
+  * moving a pod onto an under-utilized node
+  * moving a pod onto a node that meets more of the pod's affinity/anti-affinity preferences
+* moving a running pod off of a node in anticipation of a known or speculated future event
+  * draining a node in preparation for maintenance, decomissioning, auto-scale-down, etc.
+  * "preempting" a running pod to make room for a pending pod to schedule
+  * proactively/speculatively make room for large and/or exclusive pods to facilitate
+    fast scheduling in the future (often called "defragmentation")
+  * (note that these last two cases are the only use cases where the first-order intent
+    is to move a pod specifically for the benefit of another pod)
+* moving a running pod off of a node from which it is receiving poor service
+  * anomalous crashlooping or other mysterious incompatiblity between the pod and the node
+  * repeated out-of-resource killing (see #18724)
+  * repeated attempts by the scheduler to schedule the pod onto some node, but it is
+    rejected by Kubelet admission control due to incomplete scheduler knowledge
+  * poor performance due to interference from other containers on the node (CPU hogs,
+    cache thrashers, etc.) (note that in this case there is a choice of moving the victim
+    or the aggressor)
+
+## Some axes of the design space
+
+Among the key design decisions are
+* how does a pod specify its tolerance for these system-generated disruptions, and how
+  does the system enforce such disruption limits
+* for each use case, where is the decision made about when and which pods to reschedule
+  (controllers, schedulers, an entirely new component e.g. "rescheduler", etc.)
+* rescheduler design issues: how much does a rescheduler need to know about pods'
+  schedulers' policies, how does the rescheduler specify its rescheduling
+  requests/decisions (e.g. just as an eviction, an eviction with a hint about where to
+  reschedule, or as an eviction paired with a specific binding), how does the system
+  implement these requests, does the rescheduler take into account the second-order
+  effects of decisions (e.g. whether an evicted pod will reschedule, will cause
+  a preemption when it reschedules, etc.), does the rescheduler execute multi-step plans
+  (e.g. evict two pods at the same time with the intent of moving one into the space
+  vacated by the other, or even more complex plans)
+
+Additional musings on the rescheduling design space can be found [here](rescheduler.md).
+
+## Design proposal
+
+The key mechanisms and components of the proposed design are priority, preemption,
```

Comment:
I debated whether to split priority/preemption into a separate doc. FWIW the reasons I made this one doc were
* disruption budget is part of both preemption and rescheduler
* /evict subresource is part of both preemption and rescheduler
* preemption can be seen as a form of rescheduling, that is driven by the scheduler (in this proposal) rather than the rescheduler. You could imagine an alternative implementation of preemption that is driven by the rescheduler (it notices a pending pod is blocked, and moves some pod(s) out of the way so that it can schedule). I think the ideas are connected enough to warrant putting them into the same doc. It does make the doc a bit harder to digest overall, though.